### PR TITLE
test: produce chunks in resharding v3 test

### PR DESCRIPTION
**Pull request by Longarithm**
Improve the test by checking that all chunks are included, and the chain passes through one epoch after resharding.

For that, I need to properly skip chunk validation if resharding happened in the middle, as this is not a part of early MVP. One more useful and necessary change: on iterating over block range of state transition, I need to move from shard_id to prev_shard_id, so that we query proper chunk in a block.

Checked offline that without validation skip, chunks for shards 1-2 are not included.

### CODE REVIEW

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 1]
```diff
@@ -40,13 +40,19 @@ use std::time::Instant;
 pub enum MainTransition {
     Genesis { chunk_extra: ChunkExtra, block_hash: CryptoHash, shard_id: ShardId },
     NewChunk(NewChunkData),
+    // TODO(#11881): this is temporary indicator that resharding happened in the
+    // state transition covered by state witness. Won't happen in production
+    // until resharding release. Must be removed and replaced with proper
+    // resharding state transition validation.
+    ShardLayoutChange,
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:47]: My thinking was that the resharding would be in addition to the existing MainTransition. I think we will first need to apply the chunk (if present) and only then validate resharding.

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 2]
```diff
@@ -130,16 +130,30 @@ pub fn pre_validate_chunk_state_witness(
     // Blocks from the last last new chunk (exclusive) to the last new chunk (inclusive).
     let mut blocks_after_last_last_chunk = Vec::new();
 
-    {
+    let (last_chunk_shard_id, last_chunk_shard_index) = {
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:133]: nit: Can this be moved to a helper method?

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 3]
```diff
@@ -130,16 +130,30 @@ pub fn pre_validate_chunk_state_witness(
...
+        let mut last_chunk_shard_id = current_shard_id;
+        let mut last_chunk_shard_index = current_shard_index;
         loop {
             let block = store.get_block(&block_hash)?;
+            (current_shard_id, current_shard_index) =
+                epoch_manager.get_prev_shard_id(&block_hash, current_shard_id)?;
+
             let chunks = block.chunks();
-            let Some(chunk) = chunks.get(shard_index) else {
+            let Some(chunk) = chunks.get(current_shard_index) else {
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:153]: Is that off by one block? 
the shard_id and shard_index are from the prev block (that's what get_prev_shard_id gives)
the block.chunks() are from the current block

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 4]
```diff
@@ -157,12 +173,27 @@ pub fn pre_validate_chunk_state_witness(
                 break;
             }
         }
+        (last_chunk_shard_id, last_chunk_shard_index)
+    };
+
+    let last_chunk_block = blocks_after_last_last_chunk.first().ok_or_else(|| {
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:179]: sanity check for myself 
- the blocks in blocks_after_last_last_chunk are in reverse chronological order 
- the blocks include the last chunk block (it's inclusive)

File: chain/chain/src/stateless_validation/chunk_validation.rs (hunk 1 above)
Comment (wacban) [reply chain/chain/src/stateless_validation/chunk_validation.rs:47]: I guess I mean that we can have both new chunk and resharding at the same time

Comment (wacban) [reply chain/chain/src/stateless_validation/chunk_validation.rs:47]: Is that what you mean by temporary? If so I'm happy to approve to unblock the progress, but still I'm curious about your thought of the proper structure for this logic.

Comment (Longarithm) [reply chain/chain/src/stateless_validation/chunk_validation.rs:48]: Yes, this is very temporary. It is just the easiest way to skip everything related to chunk validation in our case.

File: chain/chain/src/stateless_validation/chunk_validation.rs (hunk 2 above)
Comment (Longarithm) [reply chain/chain/src/stateless_validation/chunk_validation.rs]: Done!

File: chain/chain/src/stateless_validation/chunk_validation.rs (hunk 3 above)
Comment (Longarithm) [reply chain/chain/src/stateless_validation/chunk_validation.rs]: I think it was correct. However, I made it more intuitive - by changing `current_shard_id` simultaneously with going to prev block.

File: chain/chain/src/stateless_validation/chunk_validation.rs [hunk 5]
```diff
@@ -104,6 +112,74 @@ pub fn validate_prepared_transactions(
...
+    /// (inclusive).
+    blocks_after_last_last_chunk: Vec<Block>,
+    last_chunk_shard_id: ShardId,
+    last_chunk_shard_index: ShardIndex,
+}
+
+fn get_state_witness_block_range(
+    state_witness: &ChunkStateWitness,
+    store: &ChainStore,
+    epoch_manager: &dyn EpochManagerAdapter,
```
Comment (wacban) [inline chain/chain/src/stateless_validation/chunk_validation.rs:129]: min nit: It's nice to keep the arguments from most general to most specific. In this case I would make it 
store,
epoch manager,
state witness,

It's nice to have a convention and this one is as good as any I guess.

**Review by wacban — APPROVED** [review]
LGTM, nice

**Pull request merged**